mis_builder_budget_contrib/models/mis_budget_by_account_item.py

```diff
@@ -0,0 +1,49 @@
+# Copyright 2022 Camptocamp SA (https://www.camptocamp.com)
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+
+from odoo import api, fields, models
+
+
+class MisBudgetByAccountItem(models.Model):
+
+    _inherit = "mis.budget.by.account.item"
+    _order = "budget_id, date_from, account_id, product_id"
+
+    product_id = fields.Many2one(
+        comodel_name="product.product",
+        domain="['|', ('company_id', '=', False), ('company_id', '=', company_id)]",
+    )
+
+    def _prepare_overlap_domain(self):
+        # Two budget items are not overlapping if they have different products. However,
+        # if they are similar for every other point, and they have the same product, or
+        # at least one of the lines has no defined product (which means that it includes
+        # the product of the other line), they are overlapping.
+        domain = super()._prepare_overlap_domain()
+        if self.product_id:
+            # If the current line has a product, we have overlapping for another line with
+            # the same product or an empty one
+            domain.extend(
+                [
+                    "|",
+                    ("product_id", "=", False),
+                    ("product_id", "=", self.product_id.id),
+                ]
+            )
+        # If the current line has no product, it includes all of them, so a similar line
+        # implies overlapping, no matter the value of product_id
+        return domain
+
+    @api.constrains(
+        "date_range_id",
+        "date_from",
+        "date_to",
+        "budget_id",
+        "analytic_account_id",
+        "analytic_tag_ids",
+        "account_id",
+        "product_id",
+    )
+    def _check_dates(self):
+        super()._check_dates()
+        return
```

Comment:
```suggestion
    @api.constrains("product_id")
    def _check_product(self):
        """Check dates if product changes."""
        self._check_dates()
        return
```

Reply:
If I do it, would the method be call if th analytic_account_id (for example) is modified ?

Reply:
What's your goal? Trigger `_check_dates` when the product is touched, correct?
Then this way you don't alter the original method at all and you avoid replicating all the constrained fields which might change somewhere else.